### PR TITLE
external: Improve the rgw creation function in external script

### LINKED ISSUE #11770: Improve the rgw creation function in external script
Following is a refactoring proposal:

**Note1**: in your original logic it seems that the user info was only modified when you are creating the user but not when the user already exists. Following code does the modification regardless of that. If that's the case plz update the logic accordingly.

**Note2:** 
The following code can be refactorized as it contains a lot of blocks like:

```
mycmd
try:
    return subprocess.check_output(cmd, stderr=subprocess.PIPE)
except subprocess.CalledProcessError as execErr:
    log_my_error(execErr)
```
further recatoring can be done by creating a function to run the commands.. didn't want to go further to keep the logic as close as possible to the original one.

```
    def create_rgw_user(self):
        cmd = [
            "radosgw-admin",
            "user",
            "create",
            "--uid",
            self.EXTERNAL_RGW_ADMIN_OPS_USER_NAME,
            "--display-name",
            "Rook RGW Admin Ops user",
            "--caps",
            "buckets=*;users=*;usage=read;metadata=read;zone=read",
        ]
        if self._arg_parser.dry_run:
            return self.dry_run("ceph " + " ".join(cmd))
        try:
            return subprocess.check_output(cmd, stderr=subprocess.PIPE)
        except subprocess.CalledProcessError as execErr:
            # if the user already exists, we just query it
            if execErr.returncode == errno.EEXIST:
                return None
            else:
                err_msg = (
                    "failed to execute command %s. Output: %s. Code: %s. Error: %s"
                    % (cmd, execErr.output, execErr.returncode, execErr.stderr)
                )
                raise Exception(err_msg)

    def get_rgw_user_info(self):
        cmd = [
            "radosgw-admin",
            "user",
            "info",
            "--uid",
            self.EXTERNAL_RGW_ADMIN_OPS_USER_NAME,
        ]
        try:
            return subprocess.check_output(cmd, stderr=subprocess.PIPE)
        except subprocess.CalledProcessError as execErr:
            err_msg = (
                "failed to execute command %s. Output: %s. Code: %s. Error: %s"
                % (cmd, execErr.output, execErr.returncode, execErr.stderr)
            )
            raise Exception(err_msg)

    def modify_rgw_user_info(self):
        cmd = [
            "radosgw-admin",
            "user",
            "modify",
            "--uid",
            self.EXTERNAL_RGW_ADMIN_OPS_USER_NAME,
            "--realm-new-name",
            self._arg_parser.rgw_realm_name,
        ]
        try:
            return subprocess.check_output(cmd, stderr=subprocess.PIPE)
        except subprocess.CalledProcessError as execErr:
            err_msg = (
                "failed to execute command %s. Output: %s. Code: %s. Error: %s"
                % (cmd, execErr.output, execErr.returncode, execErr.stderr)
            )
            raise Exception(err_msg)

    def add_caps_info(self):
        cmd = [
            "radosgw-admin",
            "caps",
            "add",
            "--uid",
            self.EXTERNAL_RGW_ADMIN_OPS_USER_NAME,
            "--caps",
            "info=read",
        ]
        try:
            subprocess.check_output(cmd, stderr=subprocess.PIPE)
            return True
        except subprocess.CalledProcessError as execErr:
            # if the ceph version not supported for adding `info=read` cap(rgw_validation)
            if (
                "could not add caps: unable to add caps: info=read\n"
                in execErr.stderr.decode("utf-8")
                and execErr.returncode == 244
            ):
                return False
            else:
                err_msg = (
                    "failed to execute command %s. Output: %s. Code: %s. Error: %s"
                    % (cmd, execErr.output, execErr.returncode, execErr.stderr)
                )
                raise Exception(err_msg)

    def create_rgw_admin_ops_user(self):
        user_info = self.create_rgw_user() or self.get_user_info()
        if self._arg_parser.rgw_realm_name:
            user_info = self.modify_rgw_user_info()

        # if it is python2, don't check for ceph version for adding `info=read` cap(rgw_validation)
        if sys.version_info.major < 3:
            jsonoutput = json.loads(output)
            return (
                jsonoutput["keys"][0]["access_key"],
                jsonoutput["keys"][0]["secret_key"],
                False,
            )

        # separately add info=read caps for rgw-endpoint ip validation
        info_cap_supported = self.add_caps_info()

        jsonoutput = json.loads(output)
        return (
            jsonoutput["keys"][0]["access_key"],
            jsonoutput["keys"][0]["secret_key"],
            info_cap_supported,
        )

```

_Originally posted by @rkachach in https://github.com/rook/rook/pull/11584#discussion_r1108276824_

### PR DESCRIPTION
**Pull request by Sheetalpamecha**
**Description of your changes:**
The rgw code can be refactorized as it contains a lot of blocks

**Which issue is resolved by this Pull Request:**
Resolves #11770

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: deploy/examples/create-external-cluster-resources.py [hunk 1]
```diff
@@ -1219,7 +1214,30 @@ def create_rgw_admin_ops_user(self):
...
                 )
                 sys.stderr.write(err_msg)
-                return None, None, False, "-1"
+            return info_cap_supported, None
+
+    def create_rgw_admin_ops_user(self):
+        if self._arg_parser.dry_run:
+            return self.create_rgw_user()
+
+        user_info = self.create_rgw_user()
```
Comment (parth-gr) [inline deploy/examples/create-external-cluster-resources.py:1223]: I think so `get_rgw_user` should be called from here only,

TO make it more clear

**Review by parth-gr — CHANGES_REQUESTED** [review]
Still it makes it more weird how the functions look

Maybe you can get the user first and then create if it doesn't exist,

Adding info cap can be completely different function call

File: deploy/examples/create-external-cluster-resources.py (hunk 1 above)
Comment (Sheetalpamecha) [reply deploy/examples/create-external-cluster-resources.py:1223]: okay

**Comment by Sheetalpamecha** [comment]
> Still it makes it more weird how the functions look
> Maybe you can get the user first and then create if it doesn't exist,
> Adding info cap can be completely different function call

Now create_rgw_admin_ops_user() will first check for user _query_user_info() and later create it if not found _build_create_user_cmd() and _add_info_cap() will be a complete separate function. 

_parse_user_info() serves as a helper function only to avoid repetitive code.

**Comment by parth-gr** [comment]
> > Still it makes it more weird how the functions look
> > Maybe you can get the user first and then create if it doesn't exist,
> > Adding info cap can be completely different function call
> 
> Now create_rgw_admin_ops_user() will first check for user _query_user_info() and later create it if not found _build_create_user_cmd() and _add_info_cap() will be a complete separate function.
> 
> _parse_user_info() serves as a helper function only to avoid repetitive code.

Thanks for the changes,

As per this issue https://github.com/rook/rook/issues/12244
We are limited to trust on the CI, so could you self confirm the changes doesn't effect any present mode we have

**Comment by parth-gr** [comment]
@Sheetalpamecha please update to latest master, the PR is been merged https://github.com/rook/rook/pull/12267

**Comment by Sheetalpamecha** [comment]
> @Sheetalpamecha please update to latest master, the PR is been merged #12267

Done

File: deploy/examples/create-external-cluster-resources.py [hunk 2]
```diff
@@ -1122,6 +1122,41 @@ def get_ceph_dashboard_link(self):
         return json_out["dashboard"]
 
     def create_rgw_admin_ops_user(self):
+        # If the user already exists, we query it
+        result = self._query_user_info()
+        if result is not None:
+            return result
```
Comment (parth-gr) [inline deploy/examples/create-external-cluster-resources.py:1128]: CI is passing 
But I see a lot of mistakes with this,
I think we need to make CI more robust and check more cases.

1) First of all we don't have to return single argument to parent call
This is the final output
```
return (
            jsonoutput["keys"][0]["access_key"],
            jsonoutput["keys"][0]["secret_key"],
            info_cap_supported,
            "",
        )
```

2) Secondly adding of info cap can be improved,
But for now I would say, if the user already exists but it folow the condition we should call that info cap addition,

SO info cap addition would be called from the parent function only.

Please try to check each and every case,
You can use this CI for testing if needed, I fail it forcefully so we can test the changes, https://github.com/rook/rook/pull/12162

**Review by parth-gr — CHANGES_REQUESTED** [review]

File: deploy/examples/create-external-cluster-resources.py (hunk 2 above)
Comment (parth-gr) [reply deploy/examples/create-external-cluster-resources.py:1128]: @Sheetalpamecha ^^

Comment (Sheetalpamecha) [reply deploy/examples/create-external-cluster-resources.py:1128]: Hi Parth, 
I have been trying to setup the external cluster and running this locally to test and try out more cases and that's why its taking a little more time.

> You can use this CI for testing if needed, I fail it forcefully so we can test the changes, https://github.com/rook/rook/pull/12162

I couldn't push the changes here, due to access issues.

Comment (parth-gr) [reply deploy/examples/create-external-cluster-resources.py:1128]: Setting up it locally would be a difficult process, as you need a standalone cephcluster,

I just failed this CI https://github.com/rook/rook/actions/runs/5200425621/jobs/9515629185 you can test the changes here with different permutations.

>I couldn't push the changes here, due to access issues.

No problem, thanks for taking the look:)

Comment (Sheetalpamecha) [reply deploy/examples/create-external-cluster-resources.py]: Yes I have been running into quite a few issues with manual setup.
Thanks for suggesting this. I happen to test some cases on such failing CI.

PTAL at the new modified code.

**Comment by Sheetalpamecha** [comment]
@parth-gr Please review this.

**Review by parth-gr — CHANGES_REQUESTED** [review]

File: deploy/examples/create-external-cluster-resources.py [hunk 3]
```diff
@@ -1122,6 +1122,47 @@ def get_ceph_dashboard_link(self):
...
+        # If the user already exists, we query it
+        output, error = self._query_user_info()
+
+        # If user info query fails, create the user
+        if error:
+            output, error = self._build_create_user_cmd()
+            if error:
+                return None, None, False, "-1"
+
+        if sys.version_info.major < 3:
```
[the remark this answers is not in the document]
Comment (parth-gr) [reply deploy/examples/create-external-cluster-resources.py:1173]: Great

**Comment by parth-gr** [comment]
@Sheetalpamecha would you be following up for the CI test?

**Comment by Sheetalpamecha** [comment]
Yes @parth-gr I am on it.

**Pull request closed without merging**